system/lib/wasmfs/syscalls.cpp

```diff
@@ -462,6 +462,32 @@ int __syscall_openat(int dirfd, intptr_t path, int flags, ...) {
   return doOpen(path::parseParent((char*)path, dirfd), flags, mode);
 }
 
+int __syscall_mknodat(int dirfd, intptr_t path, int mode, int dev) {
+  assert(dev == 0); // TODO: support special devices
+  if (mode & S_IFDIR) {
+    return -EINVAL;
+  }
+  if (mode & S_IFIFO) {
+    return -EPERM;
+  }
+  auto flags = O_CREAT;
+  auto read = bool(mode & WASMFS_PERM_READ);
+  auto write = bool(mode & WASMFS_PERM_WRITE);
+  if (read && !write) {
+    flags |= O_RDONLY;
+  } else if (write && !read) {
+    flags |= O_WRONLY;
+  } else if (read && write) {
+    flags |= O_RDWR;
+  }
+  auto err = doOpen(path::parseParent((char*)path, dirfd), flags, mode);
+  // Return an error if there is one, or 0 on success.
+  if (err < 0) {
+    return err;
+  }
+  return 0;
```

Comment:
We need to call `close` in the success case so we don't leak the file descriptor allocated in `doOpen`. Alternatively, we could avoid needlessly allocating a file descriptor at all by moving the logic to create new files here and having `doOpen` call out to `mknodat` when it needs to create files.

Reply:
Ah, good catch, I forgot about that global state...

I looked into having doOpen call mknodat, but it's hard to make it nice. There is a lot of parsing and validation code that would end up duplicated in the two places. Instead, I added a flag to control whether to allocate an FD which is a small change.

Reply:
Now we can just return the result of `doOpen` rather than checking whether it is zero.